### ISSUE TITLE
Better sharing story

### ISSUE THREAD

**Issue by hexylena** [galaxy-genome-annotation/galaxy-tools#47]
We need a better way for users to self-organise. My problem is:

- it should be easy for users to share data
- they should be able to group themselves and share amongst the group 
- they shouldn't have to ask an admin

I guess we have some options:

- try and push this on galaxy side (+ sync in apollo)
	- but already sharing is a mess that will take years to sort out
	- so maybe not this?
- work on it in tool side
	- write some custom "join apollo group", create group, etc.
	- needs a "share organism" tool so you don't have to update it to just share.
- try and get apollo to do it
	- but not sure they have bandwidth for this

**Comment by erasche** [comment]
Also need to think about GDPR compliance :(

**Comment by abretaud** [comment]
Looks like I need this now, at least for apollo.genouest.org. I was sure it was already possible, but I think it was just a dream

I can work on a galaxy tool to grant/revoke access to user/groups. Another one maybe to create/remove from groups too.
I could make select boxes for user/groups selection probably. Would that work with GDPR, or should there only be a text box where people would need to enter by hand userid/group?

**Comment by hexylena** [comment]
I guess we have to solve the tough questions eventually. Just going to start some discussion here if it's ok :) My ideas have changed over time a bit, too.

Feel like we need a good model for inspiration, but so few things have self organising groups. I'd argue galaxy needs this (and if they had it, I would just sync to apollo.) And i'd argue that anything we develop shouldn't add tables (if possible) or need any information outside of the apollo db.

Maybe let's model off of github, where users create + manage orgs? So, somewhere:

- a user can create a group
  - should that be public?
  - should it be an option / do we care? or just private + not worry about.
- maybe they can generate an invite link
  - with a nonce? we'd have to validate them. (e.g. if generate 20 invites for diff people.)
  - Or invite specific users
    - So for this, galaxy does say it is a *public* user name. Apollo doesn't, for non-galaxy apolli. do we care?
  - invite? or just 'add to group'
- do we need permissions within groups?
- how do we identify the group owner? (ugh does this need a table or link?)

so my feeling (lately) is a select for users would work. But maybe this shouldn't be a galaxy tool.

**Comment by abretaud** [comment]
> Maybe let's model off of github, where users create + manage orgs? So, somewhere:
> 
>     * a user can create a group
>       
>       * should that be public?
>       * should it be an option / do we care? or just private + not worry about.

I'd vote for public groups, I don't think it's a very sensitive data

>     * maybe they can generate an invite link
>       
>       * with a nonce? we'd have to validate them. (e.g. if generate 20 invites for diff people.)
>       * Or invite specific users
>         
>         * So for this, galaxy does say it is a _public_ user name. Apollo doesn't, for non-galaxy apolli. do we care?
>       * invite? or just 'add to group'

I think invites would add complexity, I'd already be happy with a simple 'add to group'

>     * do we need permissions within groups?
> 
>     * how do we identify the group owner? (ugh does this need a table or link?)

Ideally I'd like to have:
- group members who can: annotate genes, export data, but not update organisms
- group owners (one is good, multiple is better) who can: do the same as members + invite/add/remove other people, update organisms, promote/demote members to owners

> 
> 
> so my feeling (lately) is a select for users would work. But maybe this shouldn't be a galaxy tool.

What do you have in mind: implementing it in Apollo? Something external?


If I understand correctly, in Apollo currently, you can already give a user or a group admin rights on an organism, which allows to :
- grant access to other users
- not grant access to other groups, nor create groups (or maybe it's a bug on my instance: the groups tab is displayed, but empty)
- update the organism (including changing the organism storage location from the ui, which is dangerous!)
- the UI is there to modify existing users or create new users (but I guess it's not authorized on the server side, did not managed to create one)

There's also the instructor role, but it's a global role which gives a kind of admin role on all organism, but no access to apollo server settings

**Comment by hexylena** [comment]
sounds good!

I'd say let's implement external if that's ok for you? (And if gallantries starts I'll help for sure). Obviously better if done in apollo but, too hard to do :p  We could just use the 'tab' feature in apollo to add this service inside?

I wrote the 'apollo permissions manager' forever ago which was admin-focused (and js only), this sounds like it could easily be a standalone interface that talks to apollo's api with admin creds?

so to summarise:

- public groups
  - users can create them when they want
  - users can add/remove organisms
- group admins manage their own groups
  - and add/remove people when needed
- permissions:
  - group owners will get 'group admin'
  - everyone else just regular member (maybe read, write, admin levels?)
  - applies to all orgs in the group, so, not complex.

**Comment by abretaud** [comment]
FYI @mboudet is back working with us (:tada: :tada:), and he has made this little app: https://github.com/genouest/permapol
It's running on https://apollo.genouest.org/permapol/ (but needs a genouest account...), and added as a [new apollo tab](https://github.com/abretaud/docker-apollo/blob/genouest/apollo-config.groovy#L137)

Currenlty users can create groups, put users in them, and grant permissions to some of their organisms (the one they created)
It's not super tested yet, there may be some bugs/things to change

**Comment by hexylena** [comment]
Awesome! That's fantastic. Can you share some screenshots? :)

**Comment by abretaud** [comment]
Yes, screenshots!
I think the navigation could be a bit smoother, but we'll works on that. Any comments welcome of course!

![permapol1](https://user-images.githubusercontent.com/238755/91853577-b1abd500-ec62-11ea-9d0d-3f7ce2fc2628.png)
![permapol2](https://user-images.githubusercontent.com/238755/91853582-b40e2f00-ec62-11ea-86c2-741f43b077dd.png)
![permapol3](https://user-images.githubusercontent.com/238755/91853587-b5d7f280-ec62-11ea-9572-ae373753db27.png)
![permapol4](https://user-images.githubusercontent.com/238755/91853594-b7a1b600-ec62-11ea-8ed6-133a06098f02.png)
![permapol5](https://user-images.githubusercontent.com/238755/91854103-8675b580-ec63-11ea-8fff-527c35997607.png)
![permapol6](https://user-images.githubusercontent.com/238755/91854056-69d97d80-ec63-11ea-8fc7-91f798c3d434.png)

**Comment by hexylena** [comment]
Wow that's just perfect!!

tir. 1. sep. 2020, 14:58 skrev Anthony Bretaudeau <notifications@github.com

**Comment by mboudet** [comment]
(The auto-completion stuff can be disabled if needed).

Among stuff I would like to add :

- Custom permissions (for now it's read/write/export for all groups)
- For now authentification works through the proxy_header, it might be good to add an option for direct login
- Better gestion of api call (for now we have to wait until apollo actually create the group, which can be quite slow...)

**Comment by abretaud** [comment]
Deployed to usegalaxy.eu now :tada:
